### PR TITLE
BYOC base

### PR DESCRIPTION
**Pull request by TheoBrigitte**
Towards: https://github.com/giantswarm/giantswarm/issues/3316

Implement context magic for resources, but still using the same azure config everywhere.

### CODE REVIEW

File: service/controller/v3/resource/deployment/resource.go [hunk 1]
```diff
@@ -27,31 +26,26 @@ const (
...
-	CloudConfig *cloudconfig.CloudConfig
-	Debugger    *debugger.Debugger
-	Logger      micrologger.Logger
+	Debugger *debugger.Debugger
+	Logger   micrologger.Logger
 
-	Azure       setting.Azure
-	AzureConfig client.AzureClientSetConfig
+	Azure           setting.Azure
+	HostAzureConfig client.AzureClientSetConfig
```
Comment (kopiczko) [inline service/controller/v3/resource/deployment/resource.go]: Is `HostAzureConfig` needed at all in this resource?

File: service/controller/v3/resource/dnsrecord/resource.go [hunk 2]
```diff
@@ -46,19 +49,19 @@ func (r *Resource) Name() string {
 }
 
 func (r *Resource) getDNSRecordSetsClient() (*dns.RecordSetsClient, error) {
-	azureClients, err := client.NewAzureClientSet(r.azureConfig)
+	azureClients, err := client.NewAzureClientSet(r.hostAzureConfig)
```
Comment (kopiczko) [inline service/controller/v3/resource/dnsrecord/resource.go]: I don't think it will work like that.
In current/desired we use guest client to get nameservers from guest DNS zones.
In update/delete we use host cluster client to create delegations.

File: service/controller/v3/resource/deployment/resource.go (hunk 1 above)
Comment (TheoBrigitte) [reply service/controller/v3/resource/deployment/resource.go]: good point, its not used anymore

File: service/controller/v3/resource/deployment/deployment.go [hunk 3]
```diff
@@ -44,34 +45,37 @@ func (r Resource) newDeployment(ctx context.Context, customObject providerv1alph
...
+		"workerSubnetCidr":        key.VnetWorkerSubnetCIDR(customObject),
+		"masterNodes":             masterNodes,
+		"workerNodes":             workerNodes,
+		"dnsZones":                key.DNSZones(customObject),
+		"hostClusterCidr":         r.azure.HostCluster.CIDR,
+		"kubernetesAPISecurePort": key.APISecurePort(customObject),
+		"masterCloudConfigData":   masterCloudConfig,
+		"workerCloudConfigData":   workerCloudConfig,
+		"vmssMSIEnabled":          r.azure.MSI.Enabled,
+		"templatesBaseURI":        baseTemplateURI(r.templateVersion),
```
Comment (kopiczko) [inline service/controller/v3/resource/deployment/deployment.go]: Why are those removed?

		"hostClusterResourceGroupName":  r.azure.HostCluster.ResourceGroup,
		"hostClusterVirtualNetworkName": r.azure.HostCluster.VirtualNetwork,

File: service/controller/v3/resource/dnsrecord/resource.go [hunk 4]
```diff
@@ -45,20 +48,29 @@ func (r *Resource) Name() string {
...
+func (r *Resource) getDNSRecordSetsGuestClient(ctx context.Context) (*dns.RecordSetsClient, error) {
+	sc, err := controllercontext.FromContext(ctx)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	return sc.AzureClientSet.DNSRecordSetsClient, nil
+}
+
+func (r *Resource) getDNSZonesClient(ctx context.Context) (*dns.ZonesClient, error) {
```
Comment (kopiczko) [inline service/controller/v3/resource/dnsrecord/resource.go]: Can you rename that one to `getDNSZonesGuestClient` to stay consistent?

Comment (TheoBrigitte) [reply service/controller/v3/resource/dnsrecord/resource.go]: done

File: service/controller/v3/resource/deployment/deployment.go (hunk 3 above)
Comment (TheoBrigitte) [reply service/controller/v3/resource/deployment/deployment.go]: done

I did also notice this does not belong in this PR

File: service/controller/v3/resource/dnsrecord/resource.go (hunk 2 above)
Comment (TheoBrigitte) [reply service/controller/v3/resource/dnsrecord/resource.go]: fixed

**Comment by TheoBrigitte** [comment]
all good ?

I will ensure e2e tests are green before merging
